source/common/network/udp_listener_impl.cc

```diff
@@ -168,13 +168,15 @@ Api::IoCallUint64Result UdpListenerImpl::send(const UdpSendData& send_data) {
       socket_.ioHandle().sendmsg(slices.begin(), num_slices, 0, *send_data.send_address_);
 
   if (send_result.ok()) {
-    buffer.drain(send_result.rc_);
+    ASSERT(send_result.rc_ == buffer.length());
     ENVOY_UDP_LOG(trace, "sendmsg sent:{} bytes", send_result.rc_);
   } else {
-    ENVOY_UDP_LOG(debug, "sendmsg failed with error {}",
-                  static_cast<int>(send_result.err_->getErrorCode()));
+    ENVOY_UDP_LOG(debug, "sendmsg failed with error {}. Ret {}",
+                  static_cast<int>(send_result.err_->getErrorCode()), send_result.rc_);
   }
 
+  buffer.drain(send_result.rc_);
```

Comment:
Can you add a comment about this line that `rc_` has been normalized to 0 in the failing case? That's not intuitive. Otherwise agreed this is a good approach and lets the caller decide what to do with an error and un-drained buffer.

Reply:
will do